Running Chunky with Java 11

same thing, java was 11, knocked down to 8, uninstalled/reinstalled openjfx in both versions...

---------------------------------\/- the stuff i get -\/-----------------

sudo update-alternatives --config java
There are 2 choices for the alternative java (providing /usr/bin/java).

  Selection    Path                                            Priority   Status
------------------------------------------------------------
  0            /usr/lib/jvm/java-11-openjdk-amd64/bin/java      1111      auto mode
  1            /usr/lib/jvm/java-11-openjdk-amd64/bin/java      1111      manual mode
* 2            /usr/lib/jvm/java-8-openjdk-amd64/jre/bin/java   1081      manual mode

Press <enter> to keep the current choice[*], or type selection number: 


BEAST:~/Downloads$  java -jar ./ChunkyLauncher.jar --verbose
/usr/lib/jvm/java-8-openjdk-amd64/jre/bin/java -Xmx1024m -Dchunky.home=/home/drblackross/.chunky -classpath /home/drblackross/.chunky/lib/chunky-core-1.4.5.jar:/home/drblackross/.chunky/lib/commons-math3-3.2.jar se.llbit.chunky.main.Chunky
Unchecked exception caused Chunky to close.
java.lang.NoClassDefFoundError: javafx/application/Application
	at java.lang.ClassLoader.defineClass1(Native Method)
	at java.lang.ClassLoader.defineClass(ClassLoader.java:763)
	at java.security.SecureClassLoader.defineClass(SecureClassLoader.java:142)
	at java.net.URLClassLoader.defineClass(URLClassLoader.java:468)
	at java.net.URLClassLoader.access$100(URLClassLoader.java:74)
	at java.net.URLClassLoader$1.run(URLClassLoader.java:369)
	at java.net.URLClassLoader$1.run(URLClassLoader.java:363)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.net.URLClassLoader.findClass(URLClassLoader.java:362)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:349)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	at se.llbit.chunky.main.Chunky.main(Chunky.java:214)
Caused by: java.lang.ClassNotFoundException: javafx.application.Application
	at java.net.URLClassLoader.findClass(URLClassLoader.java:382)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:349)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	... 13 more

Failed to launch Chunky. Command used:
/usr/lib/jvm/java-8-openjdk-amd64/jre/bin/java -Xmx1024m -Dchunky.home=/home/drblackross/.chunky -classpath /home/drblackross/.chunky/lib/chunky-core-1.4.5.jar:/home/drblackross/.chunky/lib/commons-math3-3.2.jar se.llbit.chunky.main.Chunky

BEAST:~/Downloads$ sudo apt-get install openjfx
Reading package lists... Done
Building dependency tree       
Reading state information... Done
openjfx is already the newest version (11.0.2+1-1~18.04.2).
0 upgraded, 0 newly installed, 0 to remove and 0 not upgraded.

no matter what is done i still end up at...

BEAST:~/Downloads$  java -jar ./ChunkyLauncher.jar 
Exception in thread "main" java.lang.BootstrapMethodError: java.lang.NoClassDefFoundError: javafx/stage/Stage
	at se.llbit.chunky.launcher.ChunkyLauncher.firstTimeSetup(ChunkyLauncher.java:273)
	at se.llbit.chunky.launcher.ChunkyLauncher.main(ChunkyLauncher.java:194)
Caused by: java.lang.NoClassDefFoundError: javafx/stage/Stage
	... 2 more
Caused by: java.lang.ClassNotFoundException: javafx.stage.Stage
	at java.net.URLClassLoader.findClass(URLClassLoader.java:382)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:349)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	... 2 more

BEAST:~/Downloads$ sudo apt-get install openjfx
Reading package lists... Done
Building dependency tree       
Reading state information... Done
openjfx is already the newest version (11.0.2+1-1~18.04.2). <------ i don't think is an issue 
0 upgraded, 0 newly installed, 0 to remove and 0 not upgraded.


------------------------------------------------------------------------------

i can run Chunky on windows 10,  yes... but it will only render a blue square at whatever size i choose as a output size in tiff or png lol

been using this tut, for the windows 10 system (didn't know about f3) 
https://chunky.llbit.se/your_first_render.html

my surface pro is the only windows 10 machine in my house (hell even the house is linux lol)

I haven't tried running chunky as root (but i shouldn't have too, right?).

sooooo this is where i am, i don't think its a new issue, but i'm willing to bet i am over looking something.

crashes after trying to launch

soooo just tried to launch the core file itself as a lark

@BEAST:~/Downloads$ java --module-path /home/drblackross/Downloads/javafx-sdk-11.0.2/lib/ --add-modules javafx.controls,javafx.fxml -jar /home/drblackross/.chunky/lib/chunky-core-1.4.5.jar 
Exception in thread "main" java.lang.NoClassDefFoundError: org/apache/commons/math3/util/FastMath
	at se.llbit.chunky.resources.Texture.setTexture(Texture.java:702)
	at se.llbit.chunky.resources.Texture.<init>(Texture.java:681)
	at se.llbit.chunky.resources.Texture.<init>(Texture.java:673)
	at se.llbit.chunky.resources.Texture$1.<init>(Texture.java:41)
	at se.llbit.chunky.resources.Texture.<clinit>(Texture.java:41)
	at se.llbit.chunky.resources.TexturePackLoader.<clinit>(TexturePackLoader.java:65)
	at se.llbit.chunky.main.CommandLineOptions.<init>(CommandLineOptions.java:421)
	at se.llbit.chunky.main.Chunky.main(Chunky.java:194)
Caused by: java.lang.ClassNotFoundException: org.apache.commons.math3.util.FastMath
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:583)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:178)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:521)
	... 8 more


.... :(

--- a/launcher/src/se/llbit/chunky/launcher/ChunkyLauncher.java
+++ b/launcher/src/se/llbit/chunky/launcher/ChunkyLauncher.java
@@ -25,15 +25,22 @@ import se.llbit.chunky.resources.SettingsDirectory;
 
 import javax.swing.*;
 import java.awt.*;
+import java.awt.event.MouseAdapter;
+import java.awt.event.MouseEvent;
+import java.awt.font.TextAttribute;
 import java.io.File;
 import java.io.FileNotFoundException;
 import java.io.FileOutputStream;
 import java.io.IOException;
 import java.io.PrintStream;
 import java.net.MalformedURLException;
+import java.net.URI;
+import java.net.URISyntaxException;
 import java.net.URL;
 import java.nio.channels.Channels;
 import java.nio.channels.ReadableByteChannel;
+import java.util.Arrays;
+import java.util.Map;
 import java.util.Scanner;
 import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.atomic.AtomicReference;
@@ -224,18 +231,7 @@ public class ChunkyLauncher {
         // Javafx error
         if(retryIfMissingJavafx)
           JavaFxLocator.retryWithJavafx(args);
-        String[] errorMessages = new String[]{
-          "Error: Java cannot find javafx.",
-          "If you are using a JVM for java 11 or later, " +
-                  "javafx is no longer shipped alongside and must be installed separately",
-          "If you already have javafx installed, you need to run chunky with the command:",
-          "java --module-path <path/to/javafx/lib> --add-modules javafx.controls,javafx.fxml -jar <path/to/ChunkyLauncher.jar>"
-        };
-        if(!GraphicsEnvironment.isHeadless())
-          JOptionPane.showMessageDialog(null, errorMessages, "Cannot find javafx", JOptionPane.ERROR_MESSAGE);
-        for(String message : errorMessages) {
-          System.err.println(message);
-        }
+        showJavafxError();
       }
       e.printStackTrace(System.err);
     }
@@ -355,5 +351,64 @@ public class ChunkyLauncher {
     } else {
       return String.format("%.1f %s", fSize, unit);
     }
+  }
+
+  private static void showJavafxError() {
+    String[] errorMessages = new String[]{
+            "Error: Java cannot find javaFX.",
+            "If you are using a JVM for java 11 or later, " +
+                    "javaFX is no longer shipped alongside and must be installed separately",
+            "If you already have javaFX installed, you need to run chunky with the command:",
+            "java --module-path <path/to/javaFX/lib> --add-modules javafx.controls,javafx.fxml -jar <path/to/ChunkyLauncher.jar>"
+    };
+    String faqLink = "https://chunky.lemaik.de/java11";
+    String faqMessage = "Check out this page for more information on how to use chunky with javaFX";
+    if(!GraphicsEnvironment.isHeadless()) {
+      JTextField faqLabel;
+      if(Desktop.getDesktop().isSupported(Desktop.Action.BROWSE)) {
+        faqLabel = new JTextField(faqMessage);
+        Font font = faqLabel.getFont();
+        Map attributes = font.getAttributes();
+        attributes.put(TextAttribute.UNDERLINE, TextAttribute.UNDERLINE_ON);
+        faqLabel.setFont(font.deriveFont(attributes));
+        faqLabel.setForeground(Color.BLUE.darker());
+        faqLabel.setCursor(Cursor.getPredefinedCursor(Cursor.HAND_CURSOR));
+        faqLabel.setEditable(false);
+        faqLabel.setBackground(null);
+        faqLabel.setBorder(null);
+        faqLabel.addMouseListener(new MouseAdapter() {
+          @Override
+          public void mouseClicked(MouseEvent e) {
+            try {
+              Desktop.getDesktop().browse(new URI(faqLink));
+            } catch(IOException | URISyntaxException ioException) {
+              ioException.printStackTrace();
+            }
+          }
+        });
+      } else {
+        faqLabel = new JTextField(String.format("%s: %s", faqMessage, faqLink));
+        faqLabel.setEditable(false);
+        faqLabel.setBackground(null);
+        faqLabel.setBorder(null);
+        faqLabel.setCursor(Cursor.getPredefinedCursor(Cursor.TEXT_CURSOR));
+      }
+      Object[] dialogContent = {
+              Arrays.stream(errorMessages).map(msg -> {
+                JTextField field = new JTextField(msg);
+                field.setEditable(false);
+                field.setBackground(null);
+                field.setBorder(null);
+                field.setCursor(Cursor.getPredefinedCursor(Cursor.TEXT_CURSOR));
+                return field;
+              }).toArray(),
+              faqLabel
+      };
+      JOptionPane.showMessageDialog(null, dialogContent, "Cannot find javaFX", JOptionPane.ERROR_MESSAGE);
+    }
+    for(String message : errorMessages) {
+      System.err.println(message);
+    }
+    System.err.printf("%s: %s\n", faqMessage, faqLink);
   }
 }

--- a/launcher/src/se/llbit/chunky/launcher/ChunkyLauncher.java
+++ b/launcher/src/se/llbit/chunky/launcher/ChunkyLauncher.java
@@ -89,7 +89,8 @@ public class ChunkyLauncher {
 
       if(args.length > 0) {
         mode = LaunchMode.HEADLESS;
-        for(String arg : args) {
+        for(int i = 0; i < args.length; i++) {
+          String arg = args[i];
           switch(arg) {
             case "--nolauncher":
               mode = LaunchMode.GUI;
@@ -148,9 +149,20 @@ public class ChunkyLauncher {
               return;
             case "--noRetryJavafx":
               retryIfMissingJavafx = false;
-              // if this is the only option, we want the launcher
-              if(args.length == 1)
+              // if this is the only option, with "--javaOptions" "<param>" we want the launcher
+              if(args.length == 3)
                 forceLauncher = true;
+              break;
+            case "--javaOptions":
+              if(i == args.length-1) {
+                System.err.println("--javaOptions must be followed by the options to can chunky with");
+                System.exit(1);
+              }
+              if(settings.javaOptions.isEmpty())
+                settings.javaOptions = args[i+1];
+              else if(!settings.javaOptions.contains(args[i+1]))
+                settings.javaOptions = args[i + 1] + " " + settings.javaOptions;
+              ++i;
               break;
             default:
               if(!headlessOptions.isEmpty()) {

--- a/launcher/src/se/llbit/chunky/launcher/ChunkyLauncher.java
+++ b/launcher/src/se/llbit/chunky/launcher/ChunkyLauncher.java
@@ -61,150 +61,176 @@ public class ChunkyLauncher {
   }
 
   public static void main(String[] args) throws FileNotFoundException {
-    final LauncherSettings settings = new LauncherSettings();
-    settings.load();
+    boolean retryIfMissingJavafx = true;
 
-		/*
-		 * If there are command line arguments then we assume that Chunky should run
-     * in headless mode, unless the --nolauncher command is present in which case
-		 * we strip that and start regularly, but without launcher. The --launcher
-		 * option overrides everything else and forces the launcher to appear.
-		 */
+    try {
+      final LauncherSettings settings = new LauncherSettings();
+      settings.load();
 
-    // TODO(llbit): the control flow for launching Chunky needs to be simplified.
+      /*
+       * If there are command line arguments then we assume that Chunky should run
+       * in headless mode, unless the --nolauncher command is present in which case
+       * we strip that and start regularly, but without launcher. The --launcher
+       * option overrides everything else and forces the launcher to appear.
+       */
 
-    boolean forceLauncher = false;
-    LaunchMode mode = LaunchMode.GUI;
-    String headlessOptions = "";
+      // TODO(llbit): the control flow for launching Chunky needs to be simplified.
 
-    if (args.length > 0) {
-      mode = LaunchMode.HEADLESS;
-      for (String arg : args) {
-        switch (arg) {
-          case "--nolauncher":
-            mode = LaunchMode.GUI;
-            break;
-          case "--launcher":
-            forceLauncher = true;
-            break;
-          case "--version":
-            System.out.println("Chunky Launcher " + LAUNCHER_VERSION);
-            return;
-          case "--verbose":
-            settings.verboseLauncher = true;
-            break;
-          case "--console":
-            settings.forceGuiConsole = true;
-            break;
-          case "--update":
-          case "--updateAlpha":
-            if (arg.equals("--updateAlpha")) {
-              System.out.println("Checking for Chunky alpha/snapshot updates..");
-              settings.downloadSnapshots = true;
-            } else {
-              System.out.println("Checking for updates..");
-            }
-            UpdateChecker updateThread = new UpdateChecker(settings, new UpdateListener() {
-              @Override public void updateError(String message) {
+      boolean forceLauncher = false;
+      LaunchMode mode = LaunchMode.GUI;
+      String headlessOptions = "";
+
+      if(args.length > 0) {
+        mode = LaunchMode.HEADLESS;
+        for(String arg : args) {
+          switch(arg) {
+            case "--nolauncher":
+              mode = LaunchMode.GUI;
+              break;
+            case "--launcher":
+              forceLauncher = true;
+              break;
+            case "--version":
+              System.out.println("Chunky Launcher " + LAUNCHER_VERSION);
+              return;
+            case "--verbose":
+              settings.verboseLauncher = true;
+              break;
+            case "--console":
+              settings.forceGuiConsole = true;
+              break;
+            case "--update":
+            case "--updateAlpha":
+              if(arg.equals("--updateAlpha")) {
+                System.out.println("Checking for Chunky alpha/snapshot updates..");
+                settings.downloadSnapshots = true;
+              } else {
+                System.out.println("Checking for updates..");
               }
-
-              @Override public void updateAvailable(VersionInfo latest) {
-                try {
-                  headlessCreateSettingsDirectory();
-                } catch (FileNotFoundException e) {
-                  throw new Error(e);
+              UpdateChecker updateThread = new UpdateChecker(settings, new UpdateListener() {
+                @Override
+                public void updateError(String message) {
                 }
-                System.out.println("Downloading Chunky " + latest + ":");
-                ConsoleUpdater.update(latest, settings);
-              }
 
-              @Override public void noUpdateAvailable() {
-                System.out.println("No updates found.");
-                if (!settings.downloadSnapshots) {
-                  System.out
-                      .println("Alpha/snapshot updates are disabled, enable with --updateAlpha");
+                @Override
+                public void updateAvailable(VersionInfo latest) {
+                  try {
+                    headlessCreateSettingsDirectory();
+                  } catch(FileNotFoundException e) {
+                    throw new Error(e);
+                  }
+                  System.out.println("Downloading Chunky " + latest + ":");
+                  ConsoleUpdater.update(latest, settings);
                 }
+
+                @Override
+                public void noUpdateAvailable() {
+                  System.out.println("No updates found.");
+                  if(!settings.downloadSnapshots) {
+                    System.out
+                            .println("Alpha/snapshot updates are disabled, enable with --updateAlpha");
+                  }
+                }
+              });
+              updateThread.start();
+              return;
+            case "--setup":
+              // Configure launcher settings.
+              doSetup(settings);
+              settings.save();
+              return;
+            case "--noRetryJavafx":
+              retryIfMissingJavafx = false;
+              // if this is the only option, we want the launcher
+              if(args.length == 1)
+                forceLauncher = true;
+              break;
+            default:
+              if(!headlessOptions.isEmpty()) {
+                headlessOptions += " ";
               }
-            });
-            updateThread.start();
-            return;
-          case "--setup":
-            // Configure launcher settings.
-            doSetup(settings);
-            settings.save();
-            return;
-          default:
-            if (!headlessOptions.isEmpty()) {
-              headlessOptions += " ";
-            }
-            headlessOptions += arg;
-            break;
-        }
-      }
-      if (forceLauncher) {
-        mode = LaunchMode.GUI;
-      }
-    }
-
-    ChunkyDeployer.LoggerBuilder loggerBuilder = () -> {
-      if (settings.forceGuiConsole || (!settings.headless && settings.debugConsole)) {
-        AtomicReference<DebugConsole> console = new AtomicReference<>(null);
-        CountDownLatch latch = new CountDownLatch(1);
-        ChunkyLauncherFx.withLauncher(settings, stage -> {
-          DebugConsole debugConsole = new DebugConsole(settings.closeConsoleOnExit);
-          debugConsole.show();
-          console.set(debugConsole);
-          latch.countDown();
-        });
-        try {
-          latch.await();
-        } catch (InterruptedException ignored) {
-          // Ignored.
-        }
-        return console.get();
-      } else {
-        return new ConsoleLogger();
-      }
-    };
-
-    if (mode == LaunchMode.HEADLESS) {
-      // Chunky is being run from the console, i.e. headless mode.
-      headlessCreateSettingsDirectory();
-      settings.debugConsole = true;
-      settings.headless = true;
-      settings.chunkyOptions = headlessOptions;
-      ChunkyDeployer.deploy(settings); // Install the embedded version.
-      VersionInfo version = ChunkyDeployer.resolveVersion(settings.version);
-      if (ChunkyDeployer.canLaunch(version, null, false)) {
-        int exitCode = ChunkyDeployer.launchChunky(settings, version, LaunchMode.HEADLESS,
-            ChunkyLauncher::launchFailure, loggerBuilder);
-        if (exitCode != 0) {
-          System.exit(exitCode);
-        }
-      } else {
-        System.err.println("Could not launch selected Chunky version. Try updating with --update");
-        System.exit(1);
-      }
-    } else {
-      final boolean finalForceLauncher = forceLauncher;
-      // A callback is used to decide if the launcher should be displayed after
-      // the first time setup dialog has been shown (if needed).
-      firstTimeSetup(settings, () -> {
-        ChunkyDeployer.deploy(settings); // Install the embedded version.
-
-        if (!finalForceLauncher && !settings.showLauncher) {
-          // Skip the launcher only if we can launch this version.
-          VersionInfo version = ChunkyDeployer.resolveVersion(settings.version);
-          if (ChunkyDeployer.canLaunch(version, null, false)) {
-            if (ChunkyDeployer.launchChunky(settings, version, LaunchMode.GUI,
-                ChunkyLauncher::launchFailure,
-                loggerBuilder) == 0) {
-              return false;
-            }
+              headlessOptions += arg;
+              break;
           }
         }
-        return true;
-      });
+        if(forceLauncher) {
+          mode = LaunchMode.GUI;
+        }
+      }
+
+      ChunkyDeployer.LoggerBuilder loggerBuilder = () -> {
+        if(settings.forceGuiConsole || (!settings.headless && settings.debugConsole)) {
+          AtomicReference<DebugConsole> console = new AtomicReference<>(null);
+          CountDownLatch latch = new CountDownLatch(1);
+          ChunkyLauncherFx.withLauncher(settings, stage -> {
+            DebugConsole debugConsole = new DebugConsole(settings.closeConsoleOnExit);
+            debugConsole.show();
+            console.set(debugConsole);
+            latch.countDown();
+          });
+          try {
+            latch.await();
+          } catch(InterruptedException ignored) {
+            // Ignored.
+          }
+          return console.get();
+        } else {
+          return new ConsoleLogger();
+        }
+      };
+
+      if(mode == LaunchMode.HEADLESS) {
+        // Chunky is being run from the console, i.e. headless mode.
+        headlessCreateSettingsDirectory();
+        settings.debugConsole = true;
+        settings.headless = true;
+        settings.chunkyOptions = headlessOptions;
+        ChunkyDeployer.deploy(settings); // Install the embedded version.
+        VersionInfo version = ChunkyDeployer.resolveVersion(settings.version);
+        if(ChunkyDeployer.canLaunch(version, null, false)) {
+          int exitCode = ChunkyDeployer.launchChunky(settings, version, LaunchMode.HEADLESS,
+                  ChunkyLauncher::launchFailure, loggerBuilder);
+          if(exitCode != 0) {
+            System.exit(exitCode);
+          }
+        } else {
+          System.err.println("Could not launch selected Chunky version. Try updating with --update");
+          System.exit(1);
+        }
+      } else {
+        final boolean finalForceLauncher = forceLauncher;
+        // A callback is used to decide if the launcher should be displayed after
+        // the first time setup dialog has been shown (if needed).
+        firstTimeSetup(settings, () -> {
+          ChunkyDeployer.deploy(settings); // Install the embedded version.
+
+          if(!finalForceLauncher && !settings.showLauncher) {
+            // Skip the launcher only if we can launch this version.
+            VersionInfo version = ChunkyDeployer.resolveVersion(settings.version);
+            if(ChunkyDeployer.canLaunch(version, null, false)) {
+              if(ChunkyDeployer.launchChunky(settings, version, LaunchMode.GUI,
+                      ChunkyLauncher::launchFailure,
+                      loggerBuilder) == 0) {
+                return false;
+              }
+            }
+          }
+          return true;
+        });
+      }
+    } catch(NoClassDefFoundError e) {
+      String cause = e.getMessage();
+      if(cause.contains("javafx")) {
+        // Javafx error
+        if(retryIfMissingJavafx)
+          JavaFxLocator.retryWithJavafx(args);
+        System.err.println("Error: Java cannot find javafx.");
+        System.err.println("If you are using a JVM for java 11 or later, " +
+                "javafx is no longer shipped alongside and must be installed separately");
+        System.err.println("If you already have javafx installed, you need to run chunky with the command:");
+        System.err.println("java --module-path <path/to/javafx/lib> --add-modules javafx.controls,javafx.fxml -jar <path/to/ChunkyLauncher.jar>");
+      }
+      e.printStackTrace(System.err);
     }
   }
 

--- a/launcher/src/se/llbit/chunky/launcher/ChunkyLauncher.java
+++ b/launcher/src/se/llbit/chunky/launcher/ChunkyLauncher.java
@@ -23,6 +23,8 @@ import se.llbit.chunky.launcher.ui.DebugConsole;
 import se.llbit.chunky.launcher.ui.FirstTimeSetupDialog;
 import se.llbit.chunky.resources.SettingsDirectory;
 
+import javax.swing.*;
+import java.awt.*;
 import java.io.File;
 import java.io.FileNotFoundException;
 import java.io.FileOutputStream;
@@ -73,8 +75,6 @@ public class ChunkyLauncher {
        * we strip that and start regularly, but without launcher. The --launcher
        * option overrides everything else and forces the launcher to appear.
        */
-
-      // TODO(llbit): the control flow for launching Chunky needs to be simplified.
 
       boolean forceLauncher = false;
       LaunchMode mode = LaunchMode.GUI;
@@ -224,11 +224,18 @@ public class ChunkyLauncher {
         // Javafx error
         if(retryIfMissingJavafx)
           JavaFxLocator.retryWithJavafx(args);
-        System.err.println("Error: Java cannot find javafx.");
-        System.err.println("If you are using a JVM for java 11 or later, " +
-                "javafx is no longer shipped alongside and must be installed separately");
-        System.err.println("If you already have javafx installed, you need to run chunky with the command:");
-        System.err.println("java --module-path <path/to/javafx/lib> --add-modules javafx.controls,javafx.fxml -jar <path/to/ChunkyLauncher.jar>");
+        String[] errorMessages = new String[]{
+          "Error: Java cannot find javafx.",
+          "If you are using a JVM for java 11 or later, " +
+                  "javafx is no longer shipped alongside and must be installed separately",
+          "If you already have javafx installed, you need to run chunky with the command:",
+          "java --module-path <path/to/javafx/lib> --add-modules javafx.controls,javafx.fxml -jar <path/to/ChunkyLauncher.jar>"
+        };
+        if(!GraphicsEnvironment.isHeadless())
+          JOptionPane.showMessageDialog(null, errorMessages, "Cannot find javafx", JOptionPane.ERROR_MESSAGE);
+        for(String message : errorMessages) {
+          System.err.println(message);
+        }
       }
       e.printStackTrace(System.err);
     }

--- a/launcher/src/se/llbit/chunky/launcher/JavaFxLocator.java
+++ b/launcher/src/se/llbit/chunky/launcher/JavaFxLocator.java
@@ -1,0 +1,82 @@
+package se.llbit.chunky.launcher;
+
+import java.io.File;
+import java.io.IOException;
+import java.lang.management.ManagementFactory;
+import java.net.URISyntaxException;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Arrays;
+
+public class JavaFxLocator {
+
+
+  private static final String[] javafxPathCandidates;
+
+  static {
+    String[] javafxPathCandidates1;
+    try {
+      javafxPathCandidates1 = new String[]{
+              "C:/Program Files/openjfx/lib",
+              System.getProperty("java.home", "") + File.separator + "lib", // java home
+              new File(ChunkyLauncher.class.getProtectionDomain().getCodeSource().getLocation().toURI())
+                      .toPath().getParent().toAbsolutePath().toString() + File.separator + "lib", // directory of the .jar
+              System.getProperty("user.dir", "") + File.separator + "lib", // working directory
+              "/usr/share/openjfx/lib",
+              "/usr/lib/jvm/java-11-openjdk/lib",
+              "/usr/lib/jvm/java-12-openjdk/lib",
+              "/usr/lib/jvm/java-13-openjdk/lib",
+              "/usr/lib/jvm/java-14-openjdk/lib",
+              "/usr/lib/jvm/java-15-openjdk/lib",
+      };
+    } catch(URISyntaxException e) {
+      e.printStackTrace();
+      javafxPathCandidates1 = new String[0];
+    }
+    javafxPathCandidates = javafxPathCandidates1;
+  }
+
+  private static boolean validJavafxDirectory(Path dir) {
+    return dir.toFile().exists()
+            && dir.resolve("javafx.base.jar").toFile().exists()
+            && dir.resolve("javafx.controls.jar").toFile().exists()
+            && dir.resolve("javafx.graphics.jar").toFile().exists()
+            && dir.resolve("javafx.fxml.jar").toFile().exists();
+  }
+
+  private static void runWithJavafx(Path javafxDir, String[] args) {
+    // https://stackoverflow.com/questions/4159802/how-can-i-restart-a-java-application
+    ArrayList<String> cmd = new ArrayList<>();
+    cmd.add(System.getProperty("java.home") + File.separator + "bin" + File.separator + "java");
+    cmd.addAll(ManagementFactory.getRuntimeMXBean().getInputArguments());
+    cmd.add("--module-path");
+    cmd.add(javafxDir.toAbsolutePath().toString());
+    cmd.add("--add-modules");
+    cmd.add("javafx.controls,javafx.fxml");
+
+    cmd.add("-cp");
+    cmd.add(ManagementFactory.getRuntimeMXBean().getClassPath());
+    cmd.add(ChunkyLauncher.class.getName());
+    cmd.addAll(Arrays.asList(args));
+    cmd.add("--noRetryJavafx"); // Make sure this doesn't end up as a fork bomb
+
+    ProcessBuilder builder = new ProcessBuilder(cmd);
+    builder.inheritIO();
+    try {
+      Process process = builder.start();
+      System.exit(process.waitFor());
+    } catch(IOException | InterruptedException e) {
+      e.printStackTrace();
+    }
+  }
+
+  public static void retryWithJavafx(String[] args) {
+    for(String candiate : javafxPathCandidates) {
+      Path directory = new File(candiate).toPath();
+      if(validJavafxDirectory(directory)) {
+        runWithJavafx(directory, args);
+      }
+    }
+  }
+
+}

--- a/launcher/src/se/llbit/chunky/launcher/JavaFxLocator.java
+++ b/launcher/src/se/llbit/chunky/launcher/JavaFxLocator.java
@@ -56,7 +56,15 @@ public class JavaFxLocator {
     cmd.add(ManagementFactory.getRuntimeMXBean().getClassPath());
     cmd.add(ChunkyLauncher.class.getName());
     cmd.addAll(Arrays.asList(args));
-    cmd.add("--noRetryJavafx"); // Make sure this doesn't end up as a fork bomb
+    cmd.add("--noRetryJavafx"); // Make sure this doesn't end up as a fork bomb*
+    // add the options again so the launcher can use them for chunky
+    cmd.add("--javaOptions");
+    StringBuilder javaOptions = new StringBuilder();
+    javaOptions.append("--module-path ");
+    javaOptions.append(javafxDir.toAbsolutePath().toString());
+    javaOptions.append(" --add-modules ");
+    javaOptions.append("javafx.controls,javafx.fxml");
+    cmd.add(javaOptions.toString());
 
     ProcessBuilder builder = new ProcessBuilder(cmd);
     builder.inheritIO();

--- a/launcher/src/se/llbit/chunky/launcher/JavaFxLocator.java
+++ b/launcher/src/se/llbit/chunky/launcher/JavaFxLocator.java
@@ -7,33 +7,31 @@ import java.net.URISyntaxException;
 import java.nio.file.Path;
 import java.util.ArrayList;
 import java.util.Arrays;
+import java.util.List;
 
 public class JavaFxLocator {
 
 
-  private static final String[] javafxPathCandidates;
+  private static final List<String> javafxPathCandidates;
 
   static {
-    String[] javafxPathCandidates1;
+    javafxPathCandidates = new ArrayList<>();
+    javafxPathCandidates.add("C:/Program Files/openjfx/lib");
+    javafxPathCandidates.add("/usr/share/openjfx/lib");
+    javafxPathCandidates.add("/usr/lib/jvm/java-11-openjdk/lib");
+    javafxPathCandidates.add("/usr/lib/jvm/java-12-openjdk/lib");
+    javafxPathCandidates.add("/usr/lib/jvm/java-13-openjdk/lib");
+    javafxPathCandidates.add("/usr/lib/jvm/java-14-openjdk/lib");
+    javafxPathCandidates.add("/usr/lib/jvm/java-15-openjdk/lib");
+
+    javafxPathCandidates.add(System.getProperty("java.home", "") + File.separator + "lib"); // java home
+    javafxPathCandidates.add(System.getProperty("user.dir", "") + File.separator + "lib"); // working directory
     try {
-      javafxPathCandidates1 = new String[]{
-              "C:/Program Files/openjfx/lib",
-              System.getProperty("java.home", "") + File.separator + "lib", // java home
-              new File(ChunkyLauncher.class.getProtectionDomain().getCodeSource().getLocation().toURI())
-                      .toPath().getParent().toAbsolutePath().toString() + File.separator + "lib", // directory of the .jar
-              System.getProperty("user.dir", "") + File.separator + "lib", // working directory
-              "/usr/share/openjfx/lib",
-              "/usr/lib/jvm/java-11-openjdk/lib",
-              "/usr/lib/jvm/java-12-openjdk/lib",
-              "/usr/lib/jvm/java-13-openjdk/lib",
-              "/usr/lib/jvm/java-14-openjdk/lib",
-              "/usr/lib/jvm/java-15-openjdk/lib",
-      };
+      javafxPathCandidates.add(new File(ChunkyLauncher.class.getProtectionDomain().getCodeSource().getLocation().toURI())
+              .toPath().getParent().toAbsolutePath().toString() + File.separator + "lib"); // directory of the .jar
     } catch(URISyntaxException e) {
       e.printStackTrace();
-      javafxPathCandidates1 = new String[0];
     }
-    javafxPathCandidates = javafxPathCandidates1;
   }
 
   private static boolean validJavafxDirectory(Path dir) {
@@ -47,7 +45,7 @@ public class JavaFxLocator {
   private static void runWithJavafx(Path javafxDir, String[] args) {
     // https://stackoverflow.com/questions/4159802/how-can-i-restart-a-java-application
     ArrayList<String> cmd = new ArrayList<>();
-    cmd.add(System.getProperty("java.home") + File.separator + "bin" + File.separator + "java");
+    cmd.add(JreUtil.javaCommand(""));
     cmd.addAll(ManagementFactory.getRuntimeMXBean().getInputArguments());
     cmd.add("--module-path");
     cmd.add(javafxDir.toAbsolutePath().toString());

